sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/DefaultService.java

```diff
@@ -70,116 +53,83 @@ public DefaultService(File pathToYamlSpecification) throws Exception {
-        this.serviceSpec = YAMLServiceSpecFactory.generateServiceSpec(rawServiceSpecification);
-        capabilityValidator.validate(this.serviceSpec);
-
-        init();
-        this.plans = generatePlansFromRawSpec(rawServiceSpecification);
-        register(serviceSpec, this.plans);
-        unlock();
+        this(DefaultScheduler.newBuilder(YAMLServiceSpecFactory.generateServiceSpec(rawServiceSpecification))
+                .setPlansFrom(rawServiceSpecification));
     }
 
-    public DefaultService(ServiceSpec serviceSpecification) {
-        this(serviceSpecification, Collections.emptyList());
+    public DefaultService(ServiceSpec serviceSpecification, Collection<Plan> plans) throws Exception {
+        this(DefaultScheduler.newBuilder(serviceSpecification).setPlans(plans));
     }
 
-    public DefaultService(ServiceSpec serviceSpec, Collection<Plan> plans) {
-        this.serviceSpec = serviceSpec;
-        init();
-        this.plans = plans;
-        register(serviceSpec, this.plans);
+    public DefaultService(DefaultScheduler.Builder schedulerBuilder) throws Exception {
+        this.schedulerBuilder = schedulerBuilder;
+
+        CuratorFramework curatorClient = CuratorFrameworkFactory.newClient(
+                schedulerBuilder.getServiceSpec().getZookeeperConnection(), CuratorUtils.getDefaultRetry());
+        curatorClient.start();
+
+        InterProcessMutex curatorMutex = lock(curatorClient, schedulerBuilder.getServiceSpec().getName());
+        try {
+            register();
+        } finally {
+            unlock(curatorMutex);
+            curatorClient.close();
+        }
     }
 
-    private void lock() {
-        String rootPath = CuratorUtils.toServiceRootPath(serviceSpec.getName());
+    /**
+     * Gets an exclusive lock on service-specific ZK node to ensure two schedulers aren't running simultaneously for the
+     * same service.
+     */
+    private static InterProcessMutex lock(CuratorFramework curatorClient, String serviceName) {
+        String rootPath = CuratorUtils.toServiceRootPath(serviceName);
         String lockPath = CuratorUtils.join(rootPath, LOCK_PATH);
-        curatorMutex = new InterProcessMutex(curatorClient, lockPath);
+        InterProcessMutex curatorMutex = new InterProcessMutex(curatorClient, lockPath);
 
         LOGGER.info("Acquiring ZK lock on {}...", lockPath);
+        final String failureLogMsg = String.format("Failed to acquire ZK lock on %s. " +
+                "Duplicate service named '%s', or recently restarted instance of '%s'?",
+                lockPath, serviceName, serviceName);
         try {
-            if (!curatorMutex.acquire(10, TimeUnit.SECONDS)) {
-                LOGGER.error("Failed to acquire ZK lock. Are you running another framework named {}? Exiting.",
-                        serviceSpec.getName());
-                SchedulerUtils.hardExit(SchedulerErrorCode.LOCK_UNAVAILABLE);
+            for (int i = 0; i < LOCK_ATTEMPTS; ++i) {
```

Comment:
After a scheduler kill (e.g. restart for config update) the old lock apparently isn't cleared right away and is still held on zk's end until some timeout. Without the retry I was seeing pretty consistent scheduler task failures after reconfiguration because the (first) lock attempt would fail.